README.md
# Final Project

This is a git repository for the final project for my CS Problem Solving and Solution College Course.\

Run the main_file.py file to run the program.

changelog.txt
Cole Compton, Sun Dec 3 18:15:02 2023 -0500, Started the project and created the first commit for the main bcanch.
Cole Compton, Sun Dec 3 18:17:00 2023 -0500, Created the test branch and created its first commit.
Cole Compton, Sun Dec 3 18:55:35 2023 -0500, Added the repository to github and made changes to the repository so that git commits should show my Github account email and my name.

CCheb, Sun Dec 3 19:39:49 2023 -0500, Updated the final_project_practice-1.py file - Commented out many of the graphing parts and left the graphing to be done in the end once the data has been cleaned and parsed.
CCheb, Sun Dec 3 20:22:08 2023 -0500, First pull request from ccompton6021/test_2 branch - Update to the final_project_practice-1.py file.

Cole Compton, Mon Dec 4 11:41:39 2023 -0500, Modified the README.md file and added the LICENSE and pyproject.toml files.
Cole Compton, Mon Dec 4 11:45:17 2023 -0500, Edited the README.md file to include a title.

Joseph Companion, Tue Dec 5 13:04:24 2023 -0500, Added the project_view.py file.

Cole Compton, Tue Dec 5 17:43:09 2023 -0500, Finished creating v1 of the LICENSE and pyproject.toml files.

CCheb, Wed Dec 6 20:03:35 2023 -0500, First Commit - Wrote python functions to clean and convert the audio file, along with a function to create a spectrogram.

Joseph Companion, Thu Dec 7 10:10:01 2023 -0500, Modified the project_view.py file
Joseph Companion, Thu Dec 7 10:10:40 2023 -0500, Added the project_model.py file.
Joseph Companion, Thu Dec 7 10:11:46 2023 -0500, Added the project_controller.py file.
Joseph Companion, Thu Dec 7 10:12:33 2023 -0500, Added the main_file.py file to run the program.

CCheb, Thu Dec 7 12:16:43 2023 -0500, Second Commit - Added more to the original cleaning_modeling.py file.

Cole Compton, Thu Dec 7 13:16:31 2023 -0500, Added the audio file to the main branch of the repository.
Cole Compton, Thu Dec 7 13:39:20 2023 -0500, Uploaded v1 of the changelog.txt file.

Joseph Companion, Thu Dec 7 13:23:39 2023 -0500, Merged the branch 'ccompton6021:main' into main.
Joseph Companion, Thu Dec 7 13:31:53 2023 -0500, Uploaded files and commits to the Github repository.

Cole Compton, Thu Dec 7 13:41:46 2023 -0500, Merge pull request #2 from d1am0ndb4ck375/main branch - Added new files.
Cole Compton, Thu Dec 7 13:54:48 2023 -0500, Uploaded v2 of the changelog.txt file.

CCheb, Sat Dec 9 12:12:16 2023 -0500, Third Commit - Completely revamped the code to look more pythonic.

Cole Compton, Sun Dec 10 13:34:11 2023 -0500, Merge branch 'dev_from_seb' into test.
Cole Compton, Sun Dec 10 13:52:04 2023 -0500, Adding the __pycache__ folder.
Cole Compton, Sun Dec 10 14:13:16 2023 -0500, Fixed issue where load file button would not work as intended.
Cole Compton, Sun Dec 10 15:14:09 2023 -0500, Made edits to claning_modeling.py.

Joseph Companion, Sun Dec 10 16:00:01 2023 -0500, Cole ComptonUpdate to project_view.py.
Joseph Companion, Sun Dec 10 16:02:41 2023 -0500, Cole ComptonUpdate project_model.py.
Joseph Companion, Sun Dec 10 16:04:08 2023 -0500, Cole ComptonUpdate project_controller.py.

Cole Compton, Sun Dec 10 16:06:48 2023 -0500, Merge pull request #5 from d1am0ndb4ck375/patch-3 branch - Update to the project_controller.py file.
Cole Compton, Sun Dec 10 16:07:00 2023 -0500, Merge pull request #4 from d1am0ndb4ck375/patch-2 branch - Update to the project_model.py file.
Cole Compton, Sun Dec 10 16:07:12 2023 -0500, Merge pull request #3 from d1am0ndb4ck375/patch-1 branch - Update to the project_view.py file.
Cole Compton, Sun Dec 10 16:35:17 2023 -0500, The program is now able to import the m4a audio file.
Cole Compton, Sun Dec 10 16:35:57 2023 -0500, The program is now able to import the m4a audio file.
Cole Compton, Sun Dec 10 16:38:09 2023 -0500, I think the program now works.
Cole Compton, Sun Dec 10 16:51:22 2023 -0500, Finished v2 of the changelog.txt file.
Cole Compton, Sun Dec 10 16:53:52 2023 -0500, Finished v2 of the README.md file



project_view.py
# Import statements
import tkinter as tk
from tkinter import filedialog
import matplotlib.pyplot as plt
from matplotlib.backends.backend_tkagg import FigureCanvasTkAgg
import numpy as np
from scipy.signal import spectrogram, find_peaks

def calculate_rt60(model, spectrogram_data, decay_percentage=60):
    """
    Calculate RT60 using the decay threshold method

    Parameters:
    - spectrogram_data: 2D array representing the spectrogram
    - decay_percentage: Percentage threshold for decay (default: 60%)

    Returns:
    - rt60_values: 1D array of RT60 values for each channel
    """
    # Find the maximum value in each frequency bin for each channel
    max_amplitudes = np.max(spectrogram_data, axis=1)

    # Calculate the decay threshold as a percentage of the maximum amplitude for each channel
    decay_threshold = max_amplitudes * (decay_percentage / 100.0)

    # Initialize an array to store decay indices for each channel
    decay_indices = np.zeros(spectrogram_data.shape[0], dtype=int)

    # Iterate over channels and find the time indices where the amplitude falls below the decay threshold
    for i in range(spectrogram_data.shape[0]):
        decay_indices[i] = np.argmax(spectrogram_data[i] < decay_threshold[i])

    # Calculate the RT60 values based on the time indices
    rt60_values = decay_indices / model.sample_rate

    return rt60_values

class View:
    def __init__(self, root):
        self.root = root
        self.root.title("Audio Analyzer")

        # Load button
        self.load_button = tk.Button(root, text="Load file", command=self.load_file)
        self.load_button.pack()
        
        # Analyze button
        self.analyze_button = tk.Button(root, text="Analyze", command=self.analyze_audio)
        self.analyze_button.pack()

        # Display file name
        self.file_label = tk.Label(root, text="")
        self.file_label.pack()

        # Display time value
        self.time_label = tk.Label(root, text="")
        self.time_label.pack()

        # Plot canvas
        self.figure, self.ax = plt.subplots(figsize=(8, 6))
        self.canvas = FigureCanvasTkAgg(self.figure, master=root)
        self.canvas.get_tk_widget().pack()
        plt.gcf().subplots_adjust(bottom=.15)

       

    def load_file(self, model):
        file_path = filedialog.askopenfilename(filetypes=[("Audio files", "*.wav; *.mp3; *.aac; .m4a")])
        if file_path:
            model.load_file(file_path)
            self.file_label.config(text=f"File: {model.filename}")
            self.plot_waveform()

    def plot_waveform(self, model):
        if model.audio_data is not None:
            self.ax.clear()
            time = np.arange(0, len(model.audio_data)) / model.sample_rate
            self.ax.plot(time, model.audio_data, color='b')
            self.ax.set_xlabel('Time (s)')
            self.ax.set_ylabel('Amplitude')
            self.canvas.draw()

    def analyze_audio(self, model):
        if model.audio_data is not None:
            num_channels = model.audio_data.shape[1] if len(model.audio_data.shape) > 1 else 1

            # Calculate spectrogram
            f, t, Sxx = spectrogram(model.audio_data.T, model.sample_rate, axis=0)

            # Find peak frequency
            peak_freq_index = np.unravel_index(np.argmax(Sxx), Sxx.shape)
            peak_freq = f[peak_freq_index[0]]

            # Calculate RT60 for low, mid, and high frequencies
            low_freq_mask = (f >= 20) & (f < 500)
            mid_freq_mask = (f >= 500) & (f < 2000)
            high_freq_mask = (f >= 2000) & (f <= 10000)

            rt60_low = calculate_rt60(model, Sxx[low_freq_mask, :])
            rt60_mid = calculate_rt60(model, Sxx[mid_freq_mask, :])
            rt60_high = calculate_rt60(model, Sxx[high_freq_mask, :])

            # Display results
            self.time_label.config(text=f"Duration: {len(model.audio_data) / model.sample_rate:.2f} seconds")
            self.show_peak_frequency(peak_freq, model)
            self.show_rt60_values(rt60_low, rt60_mid, rt60_high)

    def show_peak_frequency(self, peak_freq, model):
        # Display peak frequency
        self.file_label.config(text=f"File: {model.filename} | Peak Frequency: {peak_freq:.2f} Hz")

    def show_rt60_values(self, rt60_low, rt60_mid, rt60_high):
        # Calculate the mean of the arrays
        rt60_low_scalar = np.mean(rt60_low) if np.any(rt60_low) else 0.0
        rt60_mid_scalar = np.mean(rt60_mid) if np.any(rt60_mid) > 0 else 0.0
        rt60_high_scalar = np.mean(rt60_high) if np.any(rt60_high) > 0 else 0.0

        rt60_text = f"RT60 - Low: {rt60_low_scalar:.2f} s | Mid: {rt60_mid_scalar:.2f} s | High: {rt60_high_scalar:.2f} s"
        self.time_label.config(text=self.time_label.cget("text") + f"\n{rt60_text}")


project_model.py
# import statements
from scipy.io import wavfile
import numpy as np
import librosa
from pydub import AudioSegment
import eyed3

# model class
class Model:
    def __init__(self):
        self.filename = ""
        self.sample_rate = 0
        self.audio_data = None

    def load_file(self, file_path):
        # Load audio file
        self.filename = file_path
        if file_path.lower().suffix(('.wav', '.mp3', '.aac','.m4a')):
            if file_path.lower().suffix(('.mp3', '.aac','.m4a')):
                # convert files to .wav
                self.convert_to_wav(file_path)

            # Load audio using librosa for flexible format support
            self.audio_data, self.sample_rate = librosa.load(file_path, sr=None, mono=False, res_type='kaiser_fast')

            # Handle metadata
            if len(self.audio_data.shape) > 1 and self.audio_data.shape[0] > 1:
                self.audio_data = np.mean(self.audio_data, axis=0)

            self.remove_metadata(file_path)

    def convert_to_wav(self, file_path):
        sound = AudioSegment.from_file(file_path)
        wav_path = file_path[:-4] + ".wav"
        sound.export(wav_path, format="wav")

    def remove_metadata(self, file_path):
        audiofile = eyed3.load(file_path)
        if audiofile is not None and audiofile.tag:
            audiofile.tag.frame_set.clear()
            audiofile.tag.save()


project_controller.py
# import statements
from project_view import View
from project_model import Model
from tkinter import filedialog

# controller class
class Controller:
    def __init__(self, model, view):
        self.model = model
        self.view = view

        # Connect GUI buttons to controller methods
        self.view.load_button.config(command=self.load_file)
        self.view.analyze_button.config(command=self.analyze_audio)

    def run(self):
        self.view.root.mainloop()

    def load_file(self):
        file_path = filedialog.askopenfilename(filetypes=[("Audio files", "*.wav; *.mp3; *.aac; *.m4a")])
        if file_path:
            self.model.load_file(file_path)
            self.view.file_label.config(text=f"File: {self.model.filename}")
            self.view.plot_waveform(self.model)

    def analyze_audio(self):
        self.view.analyze_audio(self.model)


final_project_practice-1.py
# -*- coding: utf-8 -*-
"""final_project_practice.ipynb

Automatically generated by Colaboratory.

Original file is located at
    https://colab.research.google.com/drive/1FBuCB6xFoQB2hOcB-9vtRBXSJcYsNXBm

Needed imports:
scipy.io -> reading audio file,
numpy -> for math computations(like cmath),
matplotib.pyplot -> for graphing results
"""

# Commented out IPython magic to ensure Python compatibility.
from scipy.io import wavfile
import scipy.io
import numpy as np
import matplotlib.pyplot as plt
# %matplotlib inline

'''
Reading the file using wavfile.read
Returns tuple with first element being the sample rate (in Hz),
second being the actual audio data (what makes the waves),
and third being the type (which we ignore)

Then we create a spectrogram which returns a tuple and we assign those
values appropriately

'''
sample_rate, data = wavfile.read('16bit1chan.wav')
spectrum, freqs, t, im = plt.specgram(data, Fs=sample_rate, \
      NFFT=1024, cmap=plt.get_cmap('autumn_r'))

# simple debugg function
# uncomment pass to disable calls
def debugg(string):
  # print(string)
   pass

# Want to establish the range of frequencies that
# we are going to work with by finding the max
# mid-range frequency (the target frequency)
def find_target_frequency(freqs):
  for x in freqs:
    if x > 1000:
      break
    return x

def frequency_check():

  # Using debugg function to check intermediate calculations
  debugg(f'freqs {freqs[:5]}')
  # Assign the mid range frequency as our target_frequency
  target_frequency = find_target_frequency(freqs)
  # Check debugg again
  debugg(f'target_frequency {target_frequency}')

  # Then we find the position (index) of where target_frequency is located
  # Used np.where as a condition to find the index in freqs array
  index_of_frequency = np.where(freqs == target_frequency)[0][0]
  debugg(f'index_of_frequency {index_of_frequency}')

  # After finding the index we look for the audio data associated with it
  # It seems that spectrum, freqs, t, etc, are all connected to each other
  # when plt.specgram was done
  data_for_frequency = spectrum[index_of_frequency]
  debugg(f'data_for_frequency {data_for_frequency}')

  # After finding the mid frequency, its index position in spectrum, and data
  # we convert its data from digital signal to decibles (db)
  data_in_db_fun = 10 * np.log10(data_for_frequency)
  return data_in_db_fun

"""Finding the -5dB"""
# Assigning the decible data into data_in_db
data_in_db = frequency_check()
plt.figure()

# Plotting the data where x is the time and y is the amplitude (power) in db
# linewidth sets the thickness of the line
# alpha sets the clarity (boldness) of the line. Ranges from 0-1
# plt.plot(t, data_in_db, linewidth=1, alpha=0.7, color='#004bc6')
# plt.xlabel('Time (s)')
# plt.ylabel('Power (dB)')

# Finding the index and value of the maximum decible
# and plotting it on the graph as a point.
# The last argument specifies that a red (keyword 'r')
# circle (keyword 'o') will be plotted according to the
# left arguments
index_of_max = np.argmax(data_in_db)
value_of_max = data_in_db[index_of_max]
# plt.plot(t[index_of_max], data_in_db[index_of_max], 'ro')

# Assigning the decible data into data_in_db
data_in_db = frequency_check()
plt.figure()

# slicing the data_in_db array so that it starts on the maximum db
sliced_array = data_in_db[index_of_max:]

# Now we find the maximum value 5 dB down which sets us up to find
# Rt20 which then helps us find Rt60
value_of_max_less_5 = value_of_max - 5
debugg(sliced_array[:10])
debugg(f'Max - 5 dB = {value_of_max_less_5}')

# We are now working with the sliced array which starts on the most
# maximum db value. Now we take that array and subtract the desired db
# ammount to get the true index and db value from sliced array(not data_in_db)
# Remmember that data_in_db contains all db values including the ones
# before the maximum.

# np.asarray just converts whatever you pass it into an array
# np.argmin returns the index of the minimum value.
def find_nearest_value(array, value):
  array = np.asarray(array)
  debugg(f'array {array[:10]}')
  idx = (np.abs(array - value)).argmin()
  debugg(f'idx {idx}')
  debugg(f'array[idx] {array[idx]}')
  return array[idx]

# See how we find the true max_less_5 db value using the function
# and then find its index in data_in_db
value_of_max_less_5 = find_nearest_value(sliced_array, value_of_max_less_5)

index_of_max_less_5 = np.where(data_in_db == value_of_max_less_5)
# After all of this we plot it
# Had to remake the graph since by only plotting the index_of_max_less_5,
# only a point showed up.
# The first four lines under this were taken from box/section 6
# plt.plot(t, data_in_db, linewidth=1, alpha=0.7, color='#004bc6')
# plt.xlabel('Time (s)')
# plt.ylabel('Power (dB)')
# plt.plot(t[index_of_max], data_in_db[index_of_max], 'ro')
# plt.plot(t[index_of_max_less_5], data_in_db[index_of_max_less_5], 'yo')

# We repeat what was done with the -5dB for the -25dB
value_of_max_less_25 = value_of_max - 25

value_of_max_less_25 = find_nearest_value(sliced_array, value_of_max_less_25)
index_of_max_less_25 = np.where(data_in_db == value_of_max_less_25)

# Then we plot everything. We could just leave all the plotting to the end
# instead of repeating the plots
plt.plot(t, data_in_db, linewidth=1, alpha=0.7, color='#004bc6')
plt.xlabel('Time (s)')
plt.ylabel('Power (dB)')
plt.plot(t[index_of_max], data_in_db[index_of_max], 'ro')
plt.plot(t[index_of_max_less_5], data_in_db[index_of_max_less_5], 'yo')
plt.plot(t[index_of_max_less_25], data_in_db[index_of_max_less_25], 'go')

# rt20 is the reverberation time (time it takes for sound to die down)
# between the -5dB and -25dB
rt20 = (t[index_of_max_less_5] - t[index_of_max_less_25])[0]

# Finally we calculate Rt60 and set some grids
rt60 = 3 * rt20

plt.grid()
plt.show()

print(f'The RT60 reverb time is {round(abs(rt60), 2)} seconds')




cleaning_modeling.py
from scipy.io import wavfile
import scipy.io
import numpy as np
import matplotlib.pyplot as plt
from pathlib import Path
from pydub import AudioSegment

# Function that converts audio file to wav
# First we check if the file is not in .wav
# if no then we check if the file extension is in .m4a
# if yes then we create a .mp4 object and convert to uncleaned .wav

# if the original file was a .wav in the first place,
# we simply create a .wav object to represent it
# After all of this, we pass it to cleaning function
def convert_audio_to_wav(audio_file_path, file_name):
    file_extension = audio_file_path.suffix
    if file_extension != '.wav':
        if file_extension == '.m4a':
            audio_file = AudioSegment.from_file(audio_file_path, format='mp4')
            audio_file.export(f'{file_name}_uncleaned.wav', format='wav')
        elif file_extension == '.mp3':
            audio_file = AudioSegment.from_mp3(audio_file_path)
            audio_file.export(f'{file_name}_uncleaned.wav', format='wav')
        else:
            audio_file = AudioSegment.from_ogg(audio_file_path)
            audio_file.export(f'{file_name}_uncleaned.wav', format='wav')
        return

# Cleaning function only checks to see if there are 2 or more channels and adjusts the .wav accordingly
# we created a new audio file if the original had 2 or more channels with the added '_mono.wav'
def clean_audio(wav_audio_file, file_name, full_file_name, audio_file_path):
    channels = wav_audio_file.channels
    if channels >= 2:
        mono_wav = wav_audio_file.set_channels(1)
        mono_wav.export(f'{file_name}_mono.wav', format='wav')
        mono_wav_audio = AudioSegment.from_file(f'{file_name}_mono.wav', format='wav')
        return mono_wav_audio, f'{file_name}_mono.wav'
    else:
        # wav_audio_file.export(f'{file_name}_mono.wav', format='wav')
        mono_wav_audio = AudioSegment.from_file(audio_file_path, format='wav')
        return mono_wav_audio, full_file_name


# simple debugg function
# uncomment pass to disable calls
def debugg(string):
    print(string)
    # pass


# Now we look for three frequencies within the audio data:
# low, mid, and high frequencies.
# Want to establish the range of frequencies that
# we are going to work with by finding the max
def find_target_frequencies(freqs):
    for x in freqs:
        if x > 500:
            low_freq = x
            break
    for x in freqs:
        if x > 1000:
            mid_freq = x
            break
    for x in freqs:
        if x > 1500:
            high_freq = x
            break
    return low_freq, mid_freq, high_freq


# frequency_check takes the low, mid, and high frequencies and finds the
# audio data for each which it then converts over to db
def frequency_check(freqs, spectrum):
    # Using debugg function to check intermediate calculations
    # debugg(f'freqs {freqs[:5]}')
    # Assign the low, mid, and high frequencies
    low_freq, mid_freq, high_freq = find_target_frequencies(freqs)

    # Then we find the position (index) of where the target freqs are located
    # Used np.whereas a condition to find the index in freqs array
    index_of_low = np.where(freqs == low_freq)[0][0]
    index_of_mid = np.where(freqs == mid_freq)[0][0]
    index_of_high = np.where(freqs == high_freq)[0][0]

    # After finding the indexes we look for the audio data associated with it.
    # It seems that spectrum, freqs, t, etc, are all connected to each other
    # when plt.specgram was done
    data_for_low = spectrum[index_of_low]
    data_for_mid = spectrum[index_of_mid]
    data_for_high = spectrum[index_of_high]

    # After finding the low, mid and, high frequencies,
    # its index positions in spectrum, and data,
    # we convert their data from digital signal to decibles (db)
    data_low_db = 10 * np.log10(data_for_low)
    data_mid_db = 10 * np.log10(data_for_mid)
    data_high_db = 10 * np.log10(data_for_high)
    return data_low_db, data_mid_db, data_high_db


# Function that returns a tuple of the max value and its index
# in accordance with the db array (low, mid, and high)
def find_max_values_and_indexes(frequency_data_collection, key):
    # Finding the index and value of the maximum decible
    # for low, mid, and high
    index_of_max = np.argmax(frequency_data_collection[key]['Data in dB'])
    value_of_max = frequency_data_collection[key]['Data in dB'][index_of_max]
    data = (value_of_max, index_of_max)
    return data


# Function that creates dB array starting at the maximum dB value
def array_slicing(frequency_data_collection, key):
    value_max, index_max = frequency_data_collection[key]['Max Value and Index']
    sliced_array = frequency_data_collection[key]['Data in dB'][index_max:]
    return sliced_array

# np.asarray just converts whatever you pass it into an array
# np.argmin returns the index of the minimum value.
def find_nearest_value(array, value):
    array = np.asarray(array)
    idx = (np.abs(array - value)).argmin()
    return array[idx]


# Now we find the nearest values of the low, mid,
# and high frequencies along with their indexes in the non_sliced
# array. In this case, we have to functions for finding the -5 dB
# and -25 dB. Both work very similar
def find_5_dB(frequency_data_collection, key, sliced_arrays):
    # We are now working with the sliced array which starts on the most
    # maximum db value. Now we take that array and subtract the desired db
    # amount to get the true index and db value from sliced array(not data_in_db)
    # Remember that data_in_db contains all db values including the ones
    # before the maximum.

    # Now we find the maximum value 5 dB down which sets us up to find
    # Rt20 which then helps us find Rt60
    value_max, index_max = frequency_data_collection[key]['Max Value and Index']

    value_less5 = value_max - 5
    value_less5 = find_nearest_value(sliced_arrays[key]['Sliced Data in dB'], value_less5)
    index_less5 = np.where(frequency_data_collection[key]["Data in dB"] == value_less5)
    return value_less5, index_less5


def find_25_dB(frequency_data_collection, key, sliced_arrays):
    value_max, index_max = frequency_data_collection[key]['Max Value and Index']

    value_less25 = value_max - 25
    value_less25 = find_nearest_value(sliced_arrays[key]['Sliced Data in dB'], value_less25)
    index_less25 = np.where(frequency_data_collection[key]["Data in dB"] == value_less25)
    return value_less25, index_less25


# Function for finding the rt_60
def find_rt60(frequency_data_collection, key, time):
    max_value_less5, index_less5 = frequency_data_collection[key]['-5 dB Value and Index']
    max_value_less25, index_less25 = frequency_data_collection[key]['-25 dB Value and Index']

    # rt20 is the reverberation time (time it takes for sound to die down)
    # between the -5dB and -25dB
    rt_20 = (time[index_less5] - time[index_less25])[0]
    rt_60 = 3 * rt_20
    return rt_60


# Graphing function that takes each of the frequencies and plots their wave form
# along with the max, -5, and -25 dB values
def graph_frequencies(frequency_data_collection, key, time):
    # Grabbing the value and indexes from dictionary
    max_value, max_index = frequency_data_collection[key]['Max Value and Index']
    value_less5, index_less5 = frequency_data_collection[key]['-5 dB Value and Index']
    value_less25, index_less25 = frequency_data_collection[key]['-25 dB Value and Index']

    # figure() allows you to create multiple graph pictures.
    # Do this everytime you want to graph something separately
    plt.figure()

    # Plotting the line
    # Plotting the data where x is the time and y is the amplitude (power) in db
    # linewidth sets the thickness of the line
    # alpha sets the clarity (boldness) of the line. Ranges from 0-1
    plt.plot(time, frequency_data_collection[key]['Data in dB'], linewidth=1, alpha=0.7, color='blue')
    plt.xlabel('Time (s)')
    plt.ylabel('Power (dB')
    plt.title(f'{key} Graph')

    # Plotting the points and adding a legend
    plt.plot(time[max_index], frequency_data_collection[key]['Data in dB'][max_index],
             'ro', label='Max dB')
    plt.plot(time[index_less5], frequency_data_collection[key]['Data in dB'][index_less5],
             'yo', label='-5 dB')
    plt.plot(time[index_less25], frequency_data_collection[key]['Data in dB'][index_less25],
             'go', label='-25 dB')
    plt.legend(loc='upper right')


# Wrapper function that connects everything together

    """
    The idea behind this program is to start by creating a dictionary which will house 
    all of the frequency data from the audio file. There are three categories: Low
    Frequency, Mid Frequency, and High Frequency in which things like the data in dB, 
    max values/indexes, -5 db values/indexes, and -25 db values/indexes will be stored inside
    of one nested dictionary for easy access. No need to scramble around looking for low, mid, or
    high frequency values. 

    The functions are designed to take the dictionary along with the keys (low, mid, and high categories)
    to create and manipulate from its values. Now there is no need to hard code values for each of the categories

    This main function shows the executing logic
    """


def main():
    # Assuming we already have user input
    # Obtaining the file name with and without extension (file_name and full_file_name)
    audio_file_path = Path(r'C:\Users\amseb_7f4cpmk\Documents\Python_Files\COP2080\final_project\16bit1chan.mp3')
    file_name = audio_file_path.stem
    full_file_name = audio_file_path.name

    # Calling the convert audio function to check if the audio file is in .wav
    # The function creates a new audio file if not in .wav
    convert_audio_to_wav(audio_file_path, file_name)

    if audio_file_path.suffix == '.wav':
        wav_file = AudioSegment.from_file(audio_file_path, format='wav')
    else:
        wav_file = AudioSegment.from_file(f'{file_name}_uncleaned.wav', format='wav')
        audio_file_path = audio_file_path.parent / f'{file_name}_uncleaned.wav'
        full_file_name = audio_file_path.name

    # Grabbing the cleaned wav file along with its name so that they can be used to create a
    # spectrogram
    wav_file_mono, f_name = clean_audio(wav_file, file_name, full_file_name, audio_file_path)

    # Creating the spectrogram
    # Reading the file using wavfile.read
    # Returns tuple with first element being the sample rate (in Hz),
    # second being the actual audio data (what makes the waves),
    # and third being the type (which we ignore)
    # Then we create a spectrogram which returns a tuple and we assign those
    # values appropriately
    sample_rate, data = wavfile.read(f_name)
    spectrum, freqs, t, im = plt.specgram(data, Fs=sample_rate,
                                          NFFT=1024, cmap=plt.get_cmap('autumn_r'))

    # Creating the dictionary which houses most of the values used throughout the program.
    # We instantiate most of the keys
    frequency_data_collection = {
        'Low Frequency': {'Max Value and Index': (), '-5 dB Value and Index': (), '-25 dB Value and Index': ()},
        'Mid Frequency': {'Max Value and Index': (), '-5 dB Value and Index': (), '-25 dB Value and Index': ()},
        'High Frequency': {'Max Value and Index': (), '-5 dB Value and Index': (), '-25 dB Value and Index': ()}}

    # Finding the dB arrays for the low, mid, and high frequencies and creating new keys
    # within the data collection to represent them
    low_in_db, mid_in_db, high_in_db = frequency_check(freqs, spectrum)
    frequency_data_collection['Low Frequency']['Data in dB'] = low_in_db
    frequency_data_collection['Mid Frequency']['Data in dB'] = mid_in_db
    frequency_data_collection['High Frequency']['Data in dB'] = high_in_db

    # Loops are created so that each of the main categories in the dictionary
    # (low, mid, and high) are all iterated through. No need to repeat the same code
    # for each category. This loop calls the max values and frequencies to find the max values
    # within Data in dB
    for key in frequency_data_collection.keys():
        data = find_max_values_and_indexes(frequency_data_collection, key)
        frequency_data_collection[key]['Max Value and Index'] = data

    # slicing the db arrays so that it starts on the maximum db
    # A new dictionary is created just to house the sliced Data in dB arrays from each of
    # the categories
    sliced_arrays = {'Low Frequency': {}, 'Mid Frequency': {}, 'High Frequency': {}}
    for key in frequency_data_collection.keys():
        array = array_slicing(frequency_data_collection, key)
        sliced_arrays[key]['Sliced Data in dB'] = array

    # Now we find the maximum value 5 and 25 dB down which sets us up to find
    # Rt20 which then helps us find Rt60
    # Finding the -5 dB values
    for key in frequency_data_collection.keys():
        data = find_5_dB(frequency_data_collection, key, sliced_arrays)
        frequency_data_collection[key]['-5 dB Value and Index'] = data

    # Finding the -25 dB values
    for key in frequency_data_collection.keys():
        data = find_25_dB(frequency_data_collection, key, sliced_arrays)
        frequency_data_collection[key]['-25 dB Value and Index'] = data

    # Loop for assigning the RT 60 for each of the categories
    for key in frequency_data_collection.keys():
        rt_60 = find_rt60(frequency_data_collection, key, t)
        frequency_data_collection[key]['RT60'] = rt_60

    # Finally we take the complete data dictionary and its main keys
    # to create three graphs: Low Frequency, Mid-Frequency, and High Frequency
    for key in frequency_data_collection.keys():
        graph_frequencies(frequency_data_collection, key, t)

    # Then, we show all the graphs
    # plt.show() shows all of the figures at once
    plt.show()


# Executing the whole program
main()
